tempesta_fw/http_sticky.c

```diff
@@ -0,0 +1,740 @@
+	+ SLEN(S_CRLFCRLF)
+/*
+ * Send an HTTP 502 response to the client. It tells the client that
+ * Tempesta is unable to forward the request to the designated server.
+ */
+int
+tfw_http_send_502(TfwHttpMsg *hm)
+{
+	u_char buf[256];
+	TfwMsg *msg;
+	TfwHttpResp *resp;
+	tfw_mastate_t state = { 0 };
+	size_t len, data_len = S_502_FIXLEN;
+	TfwConnection *conn = hm->conn;
+
+	if ((resp = (TfwHttpResp *) tfw_http_msg_alloc(Conn_Srv)) == NULL) {
+		return -1;
+	}
+	msg = (TfwMsg *)resp;
+
+	if (tfw_http_msg_setup(&state, msg, data_len) != 0) {
+		tfw_http_msg_free((TfwHttpMsg *)resp);
+		return -1;
+	}
+
+	tfw_http_msg_add_data(&state, msg, S_502, SLEN(S_502));
+	tfw_http_msg_add_data(&state, msg, S_CRLF, SLEN(S_CRLF));
+
+	tfw_http_msg_add_data(&state, msg, S_F_DATE, SLEN(S_F_DATE));
+	len = tfw_http_prep_date(buf);
+	tfw_http_msg_add_data(&state, msg, buf, len);
+	tfw_http_msg_add_data(&state, msg, S_CRLF, SLEN(S_CRLF));
+
+	tfw_http_msg_add_data(&state, msg, S_F_CONTENT_LENGTH,
+					   SLEN(S_F_CONTENT_LENGTH));
+	tfw_http_msg_add_data(&state, msg, "0", 1);
+	tfw_http_msg_add_data(&state, msg, S_CRLFCRLF, SLEN(S_CRLFCRLF));
+
+	tfw_connection_send_cli(conn, (TfwMsg *)resp);
+	tfw_http_msg_free((TfwHttpMsg *)resp);
+
+	return 0;
+}
+
+static void
+tfw_http_strim(tfw_str_t *s)
+{
+	u_char *sptr, *eptr;
+
+	if (!s->len)
+		return;
+
+	eptr = s->ptr + s->len - 1;
+	while (eptr >= s->ptr && isspace(*eptr))
+		eptr--;
+	s->len = eptr - s->ptr + 1;
+
+	sptr = skip_spaces(s->ptr);
+	s->len -= sptr - s->ptr;
+	s->ptr = sptr;
+}
+
+static TfwStr *
+tfw_http_field(TfwHttpMsg *hm, const char *field, size_t len)
+{
+	int i;
+	TfwStr *hdr;
+
+	for (i = 0; i < hm->h_tbl->size; i++) {
+		hdr = &hm->h_tbl->tbl[i].field;
+		if (tfw_str_eq_cstr(hdr, field, len,
+				    TFW_STR_EQ_PREFIX | TFW_STR_EQ_CASEI)) {
+			break;
+		}
+	}
+	if (i < hm->h_tbl->size) {
+		/* DIRTY HACK TO COMPENSATE FOR PARSER BUG */
+		const tfw_str_t s_cookie = tfw_string("Cookie:");
+		if (!strncmp(field, s_cookie.ptr, s_cookie.len))
+			hdr->len += 3;
+		return hdr;
+	}
+	return NULL;
+}
+
+static int
+tfw_http_field_val(TfwHttpMsg *hm, const tfw_str_t *field, tfw_str_t *value)
+{
+	char *buf;
+	size_t len;
+	TfwStr *hdr;
+
+	if ((hdr = tfw_http_field(hm, field->ptr, field->len)) == NULL) {
+		return 0;
+	}
+	len = tfw_str_len(hdr) + 1;
+	if ((buf = tfw_pool_alloc(hm->pool, len)) == NULL) {
+		return -ENOMEM;
+	}
+	len = tfw_str_to_cstr(hdr, buf, len);
+	value->ptr = buf + field->len;
+	value->len = len - field->len;
+	tfw_http_strim(value);
+
+	return 1;
+}
+
+/*
+ * Find Tempesta sticky cookie in an HTTP message.
+ */
+static int
+tfw_http_sticky_get(TfwHttpMsg *hm, tfw_str_t *cookie)
+{
+	int ret;
+	u_char *valptr, *endptr;
+	const tfw_str_t s_field = tfw_string("Cookie:");
+	tfw_str_t value = { 0 };
+
+	/*
+	 * Find a 'Cookie:' header field in the request.
+	 * The search for Tempesta sticky cookie within the field.
+	 * Note that there can be only one "Cookie:" header field.
```

Comment:
I have found a good answer with references here: http://stackoverflow.com/questions/16305814/are-multiple-cookie-headers-allowed-in-an-http-request Perhaps I should put the RFC and section numbers in the comment.